src/libsyntax/parse/parser.rs

```diff
@@ -2908,12 +2907,29 @@ impl<'a> Parser<'a> {
                 }
             }
             None => {
-                if as_prec > min_prec && self.eat_keyword(keywords::As) {
+                if as_prec >= min_prec && self.eat_keyword(keywords::As) {
                     let rhs = self.parse_ty();
                     let _as = self.mk_expr(lhs.span.lo,
                                            rhs.span.hi,
                                            ExprCast(lhs, rhs));
                     self.parse_more_binops(_as, min_prec)
+                } else if range_prec >= min_prec
+                        && match lhs.node { ExprRange(_, _) => false, _ => true }
+                        && self.eat(&token::DotDot) {
+                    // '..' range notation, infix or postfix form
+                    // Note that we intentionally reject other range expressions on the lhs.
+                    // This makes '..1..2' invalid.
+                    // This is necessary for consistency between the prefix and postfix forms.
```

Comment:
I'm not sure what exactly the original goal was. Note that `a..b..c..d` was already a syntax error without this pull request. This special case just preserves the non-associativity of `..` while moving the implementation to `parse_more_binops`. It also fixes some inconsistencies: `a..b..` was already invalid without this pull request, but `..a..b` was valid.
I think it's a good idea to keep `..` non-associative -- it's consistent with `a < b < c`, and means we have one less thing to worry about when thinking about the ambiguities caused by the optional expressions around `..`.